source/common/quic/envoy_quic_client_stream.cc

```diff
@@ -147,7 +147,11 @@ void EnvoyQuicClientStream::encodeData(Buffer::Instance& data, bool end_stream)
       // TODO(danzh): investigate the cost of allocating one buffer per slice.
       // If it turns out to be expensive, add a new function to free data in the middle in buffer
       // interface and re-design QuicheMemSliceImpl.
-      quic_slices.emplace_back(quiche::QuicheMemSlice::InPlace(), data, slice.len_);
+      auto single_slice_buffer = std::make_unique<Buffer::OwnedImpl>();
```

Comment:
This change I would love a way to fallback in case it goes wrong. RUNTIME_GUARD isn't an option here because we want to compile out the old way internally. Would `#ifdef SOME_PREPROCESSOR` be acceptable? @alyssawilk

Reply:
do we need to compile out the old way overnight?  Can we do the usual flip-wait-deprecate?

Reply:
ah so for _functional_ changes you need to wait 6 months for operators to update and test, but for "I want to guard this in case it causes crashes but really things look fine" changes you can remove at your own pace.  Does that help any?

Reply:
As Dan mentions, we'd like to be able to change the compile-time dependencies (in particular so we no longer depend on a custom constructor in Envoy platform's QuicheMemSliceImpl).

I'm happy to leave the default behavior as-is and protect this new behavior behind a compile-time guard so everyone who doesn't care about swapping out this dependency is unaffected, but guarding it behind a runtime flag means we can't actually swap out the underlying implementation until the old way is removed.

Reply:
I'm not sure how we get a fallback and keep the code in, so I was trying to offer a less painful timeline to get both in sequence.

Reply:
Right. I don't think a _runtime_ fallback is feasible for what I want to do.

Reply:
@RyanTheOptimist FYI. I think we want to use a runtime guard with an ifdef. Something like...

```
#ifdef USE_QUICHE_MEM_SLICE_RELEASOR_API_EXCLUSIVELY
if (!RUNTIME_GUARD(use_quiche_mem_slice_releasor)) {
  quic_slices.emplace_back(quiche::QuicheMemSlice::InPlace(), data, slice.len_);
} else {
#endif
  auto single_slice_buffer = std::make_unique<Buffer::OwnedImpl>();
  // ...
#ifdef USE_QUICHE_MEM_SLICE_RELEASOR_API_EXCLUSIVELY
}
#endif
```

(Rough pseudocode. I'll send a commit along these lines later today.)